keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -131,126 +133,178 @@ For CSI, kubelet currently uses following heuristics:
 
-**How kubelet knows *if* a volume supports mounting with SELinux?**
+### Implementation Details/Notes/Constraints [optional]
 
-[As described above](#volume-mounting), any volume can be mounted with `-o context=XYZ`, as long as it's the first mount of the volume.
+#### API changes
+`PodSecurityContext` is extended with a new field `SELinuxChangePolicy`:
 
-* For mounting of block devices, it's safe to assume that `-o context=XYZ` always works - kubelet / CSI drivers mounts such a volume as whole.
-* For shared volumes, such as NFS, GlusterFS or CephFS, it's hard for kubelet to distinguish what is a volume and what is a directory on it.
-  For example, when mounting `example.com:/exports/archive/jsafrane/projects/foo`, `example.com:/exports/archive` is name of the volume (from kernel point of view) and `jsafrane/projects/foo` is path inside the volume.
-  It can still be mounted with `-o context=XYZ`, however, all subsequent mounts of `example.com:/exports/archive` must use the same `-o context=XYZ` option!
-  In Kubernetes, we do not have any way to tell if a NFS, GlusterFS or CephFS PVs are using the same volumes (NFS/Gluster/CephFS shares), and therefore kubelet cannot use `-o context=XYZ` for these volume plugins.
+```go
+// SecurityContext holds security configuration that will be applied to a container.
+// Some fields are present in both SecurityContext and PodSecurityContext.  When both
+// are set, the values in SecurityContext take precedence.
+type SecurityContext struct {
+// ...
+	// seLinuxChangePolicy defines how the container's SELinux label is applied to all volumes used by the Pod.
+	// It has no effect on nodes that do not support SELinux or when the volume does not support SELinux.
+    // Valid values are "MountReadWriteOncePod", "Recursive", and "MountAll". If not specified, "MountReadWriteOncePod" is used.
+    // Note that this field cannot be set when spec.os.name is windows.
+    // +optional
+    SELinuxChangePolicy *PodSELinuxChangePolicy
+}
 
-* For CSI drivers, kubelet has no clue if the PVs provided by the driver are independent and each of them can be mounted with a different `-o context=` value, or some (or all) PVs are in fact the same volume from kernel point of view, only a different subdirectory of it.
-    Therefore, we need CSI driver vendors to provide such information in CSIDriver object of their driver:
+// PodSELinuxChangePolicy defines how the container's SELinux label is applied to all volumes used by the Pod.
+type PodSELinuxChangePolicy string
 
-    ```go
+const (
+	// MountReadWriteOncePod mounts PersistentVolumes with access mode ReadWriteOncePod with `-o context` mount option.
+	// All other volumes are relabeled recursively by the container runtime.
+	// Mounting with `-o context` is instant and does not require the container runtime to recursively walk through the volume. 
+    SELinuxChangePolicyMountReadWriteOncePod PodSELinuxChangePolicy = "MountReadWriteOncePod"
```

Comment:
Also - for this option to work, Pod must have an explicit SELinux context specified, otherwise it will again default to recursive. 

IMO - this default though makes sense, there isn't one single word that is going to capture all the necessary information. 

------

I was thinking - how about we keep existing default and introduce a boolean field in `securityContext`. So something like:

```securityContext.useSELInuxMountOption: true|false```

So we keep existing behaviour for `ReadWriteOncePod` and other situations (if field is `nil`) but use this field if explicitly set to true.  If this is explicitly set to `false`, then I guess we will have to use recursive chcon even for RWOP volumes. But that doesn't seem so bad.

Reply:
We should have all allowed options as explicit values in the API. `nil` should mean either `true` or `false`, not something in between.

I'll add the wording about SELinux label in Pod.

Reply:
Added a very short sentence about SELinuxLabel in PodSecurityContext / SecurityContext.